WIP: mutation

tpolecat opened the pull request:
This adds support for mutation as follows:

- `RootMapping` now has a `mutation` member of type `Mutation` (a wrapper for `(Query, Env) => F[Result[(Query, Env)]]`). All constructors default this value to `Mutation.None`, which does nothing.
- A `Mutation` performs effects in `F` and yields a [possibly updated] query and env.
- The `QueryInterpreter` executes the mutation and uses the [possibly updated] query and env to construct the result cursor.
- For the doobie and Skunk mappings the updates and subsequent queries run in separate transactions, which I think is actually what we want, given the "mutate and then query" semantics in the spec. In any case _not_ doing it this way would be a significant complication so I'm calling it good for now.

Todo:

- [x] Add an database insert test that generates a new key.
- [x] Implement database tests for doobie.

It's a pretty small change, which makes me hopeful. More comments inline.

tpolecat commented on modules/core/src/main/scala/mapping.scala:
```diff
@@ -37,7 +37,7 @@ abstract class Mapping[F[_]](implicit val M: Monad[F]) extends QueryExecutor[F,
     compiler.compile(text, name, untypedVars, introspectionLevel) match {
       case Ior.Right(operation) =>
-        run(operation.query, schema.queryType, env)
+        run(operation.query, operation.rootTpe, env)
```
This was a bug.

milessabin replied:
Good spot.

tpolecat commented on modules/skunk/src/test/scala/mutation/MutationData.scala:
```diff
@@ -0,0 +1,141 @@
+              }
+          }),
+        ),
+      ),
```
The combined ergonomics of `Env` and `Result` and `F` make this a little awkward.

milessabin replied:
Agreed, but this seems like a good starting point at least.

tpolecat commented:
Hm also some test failures. I'll look at those tomorrow.

tpolecat commented:
Ok it's working now. @milessabin I'm ready for your eyes when you have a moment. I'm not super happy with the ergonomics but that seems easy to fix with more `Mutation` constructors for common cases.

milessabin commented on modules/skunk/src/test/scala/mutation/MutationData.scala:
```diff
@@ -0,0 +1,141 @@
+  ))
+}
+
+object MutationMapping extends SkunkMappingCompanion {
```
Will this carry over to the doobie backend as well?

milessabin reviewed (approved):
LGTM, although it'd be nice to have doobie support as well.

tpolecat commented:
Cool, doobie support is on my checklist above, as well as a few more tests. Will ping you again for a final looksee when those are done.

tpolecat commented on modules/sql/src/test/scala/SqlMutationSpec.scala:
```diff
@@ -0,0 +1,109 @@
+
+trait SqlMutationSchema {
+
+  val schema =
```
It occurred to me that the schema really should be with the [shared] tests. Maybe I should change the rest of them.

tpolecat commented:
Alright I added an insert case that elaborates the `Select` at runtime and moved the test suite so it's shared with both database back-ends. Any last words?

milessabin commented:
This LGTM! :shipit:
